plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,542 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/* License note: While this "glue" code to the nDPI library is GPLv2,
+ * nDPI is itself LGPLv3 which is known to be incompatible with the
+ * GPLv2. */
+
+#include "suricata-common.h"
+#include "suricata-plugin.h"
+
+#include "detect-engine-helper.h"
+#include "detect-parse.h"
+#include "flow-callbacks.h"
+#include "flow-storage.h"
+#include "output-eve.h"
+#include "thread-callbacks.h"
+#include "thread-storage.h"
+#include "util-debug.h"
+
+#include "ndpi_api.h"
+
+static ThreadStorageId thread_storage_id = { .id = -1 };
+static FlowStorageId flow_storage_id = { .id = -1 };
+static int ndpi_protocol_keyword_id = -1;
+static int ndpi_risk_keyword_id = -1;
+
+struct NdpiThreadContext {
+    struct ndpi_detection_module_struct *ndpi;
+};
+
+struct NdpiFlowContext {
+    struct ndpi_flow_struct *ndpi_flow;
+    ndpi_protocol detected_l7_protocol;
+    uint8_t detection_completed;
+};
+
+typedef struct DetectnDPIProtocolData_ {
+    ndpi_master_app_protocol l7_protocol;
+    uint8_t negated;
+} DetectnDPIProtocolData;
+
+typedef struct DetectnDPIRiskData_ {
+    ndpi_risk risk_mask; /* uint64 */
+    uint8_t negated;
+} DetectnDPIRiskData;
+
+static void ThreadStorageFree(void *ptr)
+{
+    SCLogDebug("Free'ing nDPI thread storage");
+    struct NdpiThreadContext *context = ptr;
+    ndpi_exit_detection_module(context->ndpi);
+    SCFree(context);
+}
+
+static void FlowStorageFree(void *ptr)
+{
+    struct NdpiFlowContext *ctx = ptr;
+    ndpi_flow_free(ctx->ndpi_flow);
+    SCFree(ctx);
+}
+
+static void OnFlowInit(ThreadVars *tv, Flow *f, const Packet *p, void *_data)
+{
+    struct NdpiFlowContext *flowctx = SCCalloc(1, sizeof(*flowctx));
+    if (flowctx == NULL) {
+        FatalError("Failed to allocate nDPI flow context");
+    }
+
+    flowctx->ndpi_flow = ndpi_flow_malloc(SIZEOF_FLOW_STRUCT);
+    if (flowctx->ndpi_flow == NULL) {
+        FatalError("Failed to allocate nDPI flow");
+    }
+
+    memset(flowctx->ndpi_flow, 0, SIZEOF_FLOW_STRUCT);
+    flowctx->detection_completed = 0;
+    FlowSetStorageById(f, flow_storage_id, flowctx);
+}
+
+static void OnFlowUpdate(ThreadVars *tv, Flow *f, Packet *p, void *_data)
```

Comment:
is it a problem that this gets the raw data in the order as it comes in? So there will be no stream reassembly, reordering and normalization of any of the data sent to ndpi.

Reply:
Getting data as it comes in is fine for ndpi